### ISSUE TITLE
Asp.net core DI and ServiceScope

### ISSUE THREAD

**Issue by dazinator** [HangfireIO/Hangfire#885]
I have been using Hangfire job server, in my asp.net core application.

In `startup.cs` I use a custom `Activator` so that I can inject any registered services, into jobs:

```

 public void Configure(IApplicationBuilder app)
            {

 // shortened for brevity
 GlobalConfiguration.Configuration.UseActivator(new ServiceProviderJobActivator(app.ApplicationServices));


}

```

The activator looks like this:

```

   public class ServiceProviderJobActivator : JobActivator
    {
        private IServiceProvider _serviceProvider;

        public ServiceProviderJobActivator(IServiceProvider serviceProvider)
        {
          // Service provider from IApplicationBuilder.ApplicationServices.
            _serviceProvider = serviceProvider;
        }

        public override object ActivateJob(Type type)
        {
            var implementation = serviceProvider.GetRequiredService(type);
            return implementation;

        }
    }

```

I am then injecting a `DbContext` into some of my jobs.

I have hit the issue, that the same instance of the `DbContext` was being accessed concurrently from different threads, because the same instance was being injected into multiple jobs running in concurrently : https://github.com/aspnet/EntityFramework/issues/8450#issuecomment-301117278

Two possible suggestions are:

1. Register my `DbContext` as transient. This means each hangfire job will get a new instance. But it also impacts the rest of my application running in the same process (i.e my MVC controllers will now have a transient instance rather than scoped to http context) which is not desirable.

2. Create a service "scope" at an appropriate level that ensures hangfire will only run a single thread at a time, within that same service scope.

I'd like to solve the problem using number 2. However I am not sure how to appropriately create and manage these service "scopes" for Hangfire. 

Has anyone else solved this already that would care to share a solution?

**Comment by dazinator** [comment]
I thought perhaps creating a new scope, for activating each job, might solve the issue. i.e like this:

```

  public override object ActivateJob(Type type)
        {

            var scopeFactory = _serviceProvider.GetService<IServiceScopeFactory>();
            var scope = scopeFactory.CreateScope();
            var scopedContainer = scope.ServiceProvider;           
              
            var implementation = scopedContainer.GetRequiredService(type);
            return implementation;

        }

```

But the problem here is that: I should `dispose` of the scope after the job has finished. Not sure of an appropriate hook to do that.

**Comment by dcorbett-fivebridgesllc** [comment]
Hangfire uses this to create an instance of the class on which it's going to call the method to run the job.   Shouldn't hangfire call dispose on this same class, and wouldn't that work for you?   You can implement a dispose your own dispose override in that class to ensure that your DbContext is disposed properly.

**Comment by dazinator** [comment]
Not really. 

So the issue is, when the job is activated, I need to first create a new `ServiceScope` and resolve the job from that scope, so that the `DbContext` that is injected into the job classes constructor is a scoped instance. If I don't do this, and resolve the job from the `ServiceProvider` without creating a scope, then multiple jobs running in parallel will het the same DbContext instance injected leading to errors. I understand this, however the question now becomes, if I create a new `ServiceScope` within `ActivateJob`, at what point can I Dispose() of it? 

Overriding Dispose in a job to dispose of the injected DbContext isn't necessary, as the DbContext (along with any other scoped services) will be disposed of, when the `ServiceScope` is disposed. Unfortunately, the job doesn't have access to the `ServiceScope` as you can't inject `ServiceScope` into the job, and you wouldn't want to.

I basically need `JobActivator` to dispose of the scope, after the job has finished running.

**Comment by burningice2866** [comment]
Since the scope is an implementation detail of your code and not anything Hangfire is aware of i don't see it as unreasonable for you to have the full responsibility of disposing it again.

How you would do it is tricky, since the only reference to the scope is on the second line of your method so you would somehow need to pass that as a reference to your job as well for it to dispose of when its finished. You could do it by adding a IServiceScope argument to your job-constructor and add the scope-instance to your container so Hangfire knows how to resolve it and inject into your job.

**Comment by dazinator** [comment]
@burningice2866 - Yeah it's a good suggestion and I gave this one some thought (i.e passing the scope into the job) and I agree with you it's not unreasonable to expect custom code to manage the scope, and thats what I am trying to acheive. I don't think it's unreasonable to expect an appropriate hook that allows us to manage this scope appropriately. Injecting the service scope into the job, and making the job dispose of it, on dispose I guess would works as long as:

1. The life of a service scope = the lifetime of a single job. 

I'm not sure if the above is optimal though, as I don't know enough about hangfire. It's possible that you would want to create a service scope at a higher level, and activate multiple jobs within the same service scope - as long as those jobs aren't executing concurrently on multiple threads. I don't know enough about how hangfire manages / schedules jobs to know if that's a desirable thing. For example, if each hangfire job server could only run a single job at a time, then you might want to create 1 ServiceScope for each Hangfire job server, and cache them. You wouldn't want to create one necessarily for each job and dispose of it for each job.. What do you think?

**Comment by dazinator** [comment]
Discovered global filters: GlobalJobFilters.Filters.Add(new LogEverythingAttribute());
Thinking this might be better than than trying to inject scope, and requiring job to dispose. Will give it a go.

**Comment by burningice2866** [comment]
Have you looked at [`JobActivator.BeginScope`](https://github.com/HangfireIO/Hangfire/blob/master/src/Hangfire.Core/JobActivator.cs#L57)?

It looks like Hangfire is creating an explicit scope [around each job being executed](https://github.com/HangfireIO/Hangfire/blob/master/src/Hangfire.Core/Server/CoreBackgroundJobPerformer.cs#L47) and its up to your implementation to return a scope here which Hangfire can dispose of again.

**Comment by dazinator** [comment]
Gave up trying to use a global filter to dispose of the service scope. The activator only has a Type for the job. So if it creates a scope, it can only store it under the jobs Type. Ideally it would need to be stored under an Id for the job, so that the Global Filter could dispose of the scope for the particular job ID after the job as finished executing in `OnPerformed`.

**Comment by dazinator** [comment]
I noticed that the Activator as a `BeginScope()` method. Could this be related? What is it for?

**Comment by burningice2866** [comment]
As referred to, Hangfire wraps the execution of each job in a BeginScope which is disposed at the end. The job-instance is resolved using the scope, which in turn can do its magic or just default back to calling ActicateJob on JobActivator which happens in the default SimpleJobActivatorScope implementation.

Soo, by the looks of it, you should override BeginScope and return a custom implementation of JobActivatorScope.

**Comment by dazinator** [comment]
> Soo, by the looks of it, you should override BeginScope and return a custom implementation of JobActivatorScope.

Looks good, i'll try that.

**Comment by dazinator** [comment]
Hmm.. despite me doing this:

```
  GlobalConfiguration.Configuration.UseActivator(new ServiceProviderJobActivator(app.ApplicationServices));

```

My activator isn't being called. In the stack trace from within the job's constructor,  I am seeing:

```
>	Hangfire.AspNetCore.dll!Hangfire.AspNetCore.AspNetCoreJobActivatorScope.Resolve(System.Type type)	Unknown

```

Any ideas why?

**Comment by dazinator** [comment]
Ok going to close this, as got to the bottom of it at last.

Hangfire handles this in `1.6.12` by registering a service that does what we suggested: https://github.com/HangfireIO/Hangfire/blob/129707d66fde24dc6379fb9d6b15fa0b8ca48605/src/Hangfire.AspNetCore/HangfireServiceCollectionExtensions.cs#L85

I was looking at this, because I was using `1.6.7` and this didn't exist in 1.6.7. Upgrading has sorted everything.

Although it may be a separate bug, in that if `AddHangfire()` registers the specialised Activator for asp.net core, and seemingly, then calling UseActivator<> isn't honoured. Perhaps its order dependent.

**Comment by burningice2866** [comment]
Looking at the code here https://github.com/HangfireIO/Hangfire/blob/master/src/Hangfire.Core/Obsolete/StartupConfiguration.cs#L73 i would say the order matters and last one wins.

**Comment by stajs** [comment]
Glad you have a way forward, but I gotta ask: why were you trying to dispose of your scope at all?

Our situation is probably a little different to yours. We are running a self-hosted (using Kestrel) website in a Windows Service, but we have been using a very similar job activator class to [your second comment](https://github.com/HangfireIO/Hangfire/issues/885#issuecomment-301470013) since the DNX days and we are still on Hangfire 1.6.6:

```csharp
public class HangfireJobActivator : JobActivator
{
	private readonly IServiceProvider _serviceProvider;

	public HangfireJobActivator(IServiceProvider serviceProvider)
	{
		if (serviceProvider == null)
			throw new ArgumentNullException(nameof(serviceProvider));

		_serviceProvider = serviceProvider;
	}

	public override object ActivateJob(Type jobType)
	{
		// Ensure jobs run in their own "request" scope. Please.
		var scopeFactory = _serviceProvider.GetService<IServiceScopeFactory>();
		var scope = scopeFactory.CreateScope();
		var job = scope.ServiceProvider.GetService(jobType);

		if (job == null)
			throw new Exception($"Could not activate a job of type '{jobType.FullName}'.");

		return job;
	}
}
```

We too felt the pain of concurrent `DbContext` access from concurrent jobs and this has been solid in our experience for the last ~2 years.

**Comment by dazinator** [comment]
By not disposing of the Service Scope, and just letting it go out of scope, it just means you are relying on the GC to run at some unknown point in the future and call Dispose on it for you. In theory, if the GC ran and disposed of the scope whilst a job was running and using an injected DbContext, id expect that DbContext could suddenly be disposed of whilst your job is still using it. The fact you say there are no issues may be because your jobs finish before the GC gets to disposing of the scope.   That aside, imho its still better to dispose of disposables rather than rely on the GC to do it at some point in the future imho, as you can free up resources earlier.

**Comment by stajs** [comment]
Yeah, for sure. I don't disagree with you, I just find these discussions intriguing. I wouldn't try and out guess the GC unless I could measure a problem, but I agree with the spirit of your point for the same reason you might wrap something in a `using`.

Interesting theory about the GC disposing a scope mid-job: is that possible? Naively I thought there was a reference to the scope for the life of a job, but perhaps you are right and we have never hit it.
